azure-kusto-ingest/azure/kusto/ingest/resource_manager.py

```diff
@@ -0,0 +1,94 @@
+"""This module is serve as a cache to all resources needed by the kusto ingest client."""
+
+from datetime import datetime, timedelta
+import random
+
+from .connection_string import _ConnectionString
+
+class _IngestClientResources:
+    def __init__(self,
+                 secured_ready_for_aggregation_queues = None,
+                 failed_ingestions_queues = None,
+                 successful_ingestions_queues = None,
+                 containers = None,
+                 status_table = None
+                 ):
+        self.secured_ready_for_aggregation_queues = secured_ready_for_aggregation_queues
+        self.failed_ingestions_queues = failed_ingestions_queues
+        self.successful_ingestions_queues = successful_ingestions_queues
+        self.containers = containers
+        self.status_table = status_table
+    
+    def is_applicable(self):
+        resources = [self.secured_ready_for_aggregation_queues, self.failed_ingestions_queues, self.failed_ingestions_queues, self.containers, self.status_table]
+        return all(resources)
+
+class _ResourceManager:
+    def __init__(self, 
+                 kusto_client):
+        self._kusto_client = kusto_client
+        self._refresh_period = timedelta(hours=1)
+
+        self._ingest_client_resources = None
+        self._ingest_client_resources_last_update = None
+
+        self._authorization_context = None
+        self._authorization_context_last_update = None
+
+    def _refresh_ingest_client_resources(self):
+        if not self._ingest_client_resources \
+           or (self._ingest_client_resources_last_update + self._refresh_period) <= datetime.now() \
+           or not self._ingest_client_resources.is_applicable():
+                self._ingest_client_resources = self._get_ingest_client_resources_from_service()
+                self._ingest_client_resources_last_update = datetime.now()  
+
+    def _get_resource_by_name(self, df, resource_name):
+        resource = df[df['ResourceTypeName'] == resource_name].StorageRoot.map(_ConnectionString.parse).tolist()
+        return resource
+    
+    def _get_ingest_client_resources_from_service(self):
+        df = self._kusto_client.execute("NetDefaultDB", ".get ingestion resources").to_dataframe()
+
+        secured_ready_for_aggregation_queues = self._get_resource_by_name(df, 'SecuredReadyForAggregationQueue')
+        failed_ingestions_queues = self._get_resource_by_name(df, 'FailedIngestionsQueue')
+        successful_ingestions_queues = self._get_resource_by_name(df, 'SuccessfulIngestionsQueue')
+        containers = self._get_resource_by_name(df, 'TempStorage')
+        status_table = self._get_resource_by_name(df, 'IngestionsStatusTable')[0]
```

Comment:
Let's keep the IngestionStatusTable generic enough so it can have more than item. So we won't forbid our customers to upgrade later

Reply:
Sounds fair enough.